index.js

```diff
@@ -0,0 +1,90 @@
+'use strict'
+
+const fp = require('fastify-plugin')
+const CSRF = require('csrf')
+const { Forbidden } = require('http-errors')
+
+async function csrfPlugin (fastify, opts) {
+  const tokens = new CSRF()
+
+  const cookieKey = opts.cookieKey || '_csrf'
+  const cookieOpts = opts.cookieOpts || { path: '/', sameSite: true }
+  const sessionKey = opts.sessionKey || '_csrf'
+  const getToken = opts.getToken || getTokenDefault
+  const sessionPlugin = opts.sessionPlugin || null
+
+  if (sessionPlugin === 'fastify-secure-session') {
+    fastify.decorateReply('generateCsrf', generateCsrfSecureSession)
+  } else if (sessionPlugin === 'fastify-session') {
+    fastify.decorateReply('generateCsrf', generateCsrfSession)
+  } else {
+    fastify.decorateReply('generateCsrf', generateCsrfCookie)
+  }
```

Comment:
Why do these need to be specific? Why not require the registered session plugin to conform to a specific protocol. It looks to me like the only difference is that `fastify-secure-session` needs to support indexed retrieval on the session object. In fact, it's the only odd man out. https://github.com/jsumners/fastify-server-session also uses indexed retrieval.

Reply:
Generally, I would agree with you, but it would require a lot of additional work which I don't have time to do at the moment. (side note, I prefer the current `fastify-secure-session` api).
I've used this approach as I very similar to what we did with `point-of-view`, where you need to specify the view engine so `point-of-view` knows what to do.